flow/MkCert.h

```diff
@@ -0,0 +1,171 @@
+
+struct CertKind {
+
+	CertKind() noexcept = default;
+
+	template <class Kind>
+	CertKind(Kind kind) noexcept : value(std::in_place_type<Kind>, kind) {}
+
+	template <class Kind>
+	bool is() const noexcept {
+		return std::holds_alternative<Kind>(value);
+	}
+
+	template <class Kind>
+	Kind const& get() const {
+		return std::get<Kind>(value);
+	}
+
+	bool isServerSide() const noexcept { return is<ServerRootCA>() || is<ServerIntermediateCA>() || is<Server>(); }
+
+	bool isClientSide() const noexcept { return !isServerSide(); }
+
+	bool isRootCA() const noexcept { return is<ServerRootCA>() || is<ClientRootCA>(); }
+
+	bool isIntermediateCA() const noexcept { return is<ServerIntermediateCA>() || is<ClientIntermediateCA>(); }
+
+	bool isLeaf() const noexcept { return is<Server>() || is<Client>(); }
+
+	bool isCA() const noexcept { return !isLeaf(); }
+
+	StringRef getCommonName(StringRef prefix, Arena& arena) const {
+		auto const side = std::string(isClientSide() ? " Client" : " Server");
+		if (isIntermediateCA()) {
+			auto const level = isClientSide() ? get<ClientIntermediateCA>().level : get<ServerIntermediateCA>().level;
+			return prefix.withSuffix(fmt::format("{} Intermediate {}", side, level), arena);
+		} else if (isRootCA()) {
+			return prefix.withSuffix(fmt::format("{} Root", side), arena);
+		} else {
+			return prefix.withSuffix(side, arena);
+		}
+	}
+
+	std::variant<ServerRootCA, ServerIntermediateCA, Server, ClientRootCA, ClientIntermediateCA, Client> value;
+};
+
+struct CertSpecRef {
+	using SelfType = CertSpecRef;
+	long serialNumber;
+	// offset in number of seconds relative to now, i.e. cert creation
+	long offsetNotBefore;
+	long offsetNotAfter;
+	VectorRef<Asn1EntryRef> subjectName;
+	// time offset relative to time of cert creation (now)
+	VectorRef<Asn1EntryRef> extensions;
+	// make test-only sample certificate whose fields are inferred from CertKind
+	static SelfType make(Arena& arena, CertKind kind);
+};
+
+struct CertAndKeyRef {
+	using SelfType = CertAndKeyRef;
+	StringRef certPem;
+	StringRef privateKeyPem;
+
+	void printCert(FILE* out) {
+		if (!certPem.empty()) {
+			::mkcert::printCert(out, certPem);
+		}
+	}
+
+	void printPrivateKey(FILE* out) {
+		if (!privateKeyPem.empty()) {
+			::mkcert::printPrivateKey(out, privateKeyPem);
+		}
+	}
+
+	bool empty() const noexcept { return certPem.empty() && privateKeyPem.empty(); }
+
+	SelfType deepCopy(Arena& arena) {
+		auto ret = SelfType{};
+		if (!certPem.empty())
+			ret.certPem = StringRef(arena, certPem);
+		if (!privateKeyPem.empty())
+			ret.privateKeyPem = StringRef(arena, privateKeyPem);
+		return ret;
+	}
+
+	// Empty (default) issuer produces a self-signed certificate
+	static SelfType make(Arena& arena, CertSpecRef spec, CertAndKeyRef issuer);
+};
+
+using CertChainRef = VectorRef<CertAndKeyRef>;
+
+// Concatenate chain of PEMs to one StringRef
+StringRef concatCertChain(Arena& arena, CertChainRef chain);
+
+enum class ESide : int { Server, Client };
```

Comment:
It stands for enumeration. I read it from some of the files in FDB codebase and thought it might be a good way to distinguish between regular OOP classes and enum classes: e.g. flow/network.h fdbclient/BackupAgent.actor.h fdbclient/JsonBuilder.h fdbrpc/AsyncFileKAIO.actor.h fdbserver/VersionedBTree.actor.cpp

Reply:
I see. A good idea